File: sdk/python/zeroid/did/document.py
"""W3C DID Document dataclasses.

Implements the core DID Document data model per the W3C DID Core specification
(https://www.w3.org/TR/did-core/).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class DIDStatus(Enum):
    """Status of a DID."""

    ACTIVE = "active"
    DEACTIVATED = "deactivated"
    SUSPENDED = "suspended"


@dataclass(frozen=True)
class VerificationMethod:
    """A verification method within a DID Document.

    Attributes:
        id: Fully qualified verification method ID.
        type: Cryptographic suite type.
        controller: DID of the controller.
        public_key_hex: Hex-encoded public key.
    """

    id: str
    type: str
    controller: str
    public_key_hex: str


@dataclass(frozen=True)
class ServiceEndpoint:
    """A service endpoint within a DID Document.

    Attributes:
        id: Fully qualified service ID.
        type: Service type identifier.
        endpoint: Service endpoint URI.
    """

    id: str
    type: str
    endpoint: str


@dataclass
class DIDDocument:
    """A W3C-compliant DID Document.

    Attributes:
        id: The DID this document describes.
        context: JSON-LD context URIs.
        verification_methods: List of verification methods.
        authentication: List of verification method IDs for authentication.
        assertion_method: List of verification method IDs for assertions.
        services: List of service endpoints.
        status: Current status of the DID.
        created: ISO 8601 creation timestamp.
        updated: ISO 8601 last-update timestamp.
    """

    id: str
    context: list[str] = field(
        default_factory=lambda: ["https://www.w3.org/ns/did/v1"]
    )
    verification_methods: list[VerificationMethod] = field(default_factory=list)
    authentication: list[str] = field(default_factory=list)
    assertion_method: list[str] = field(default_factory=list)
    services: list[ServiceEndpoint] = field(default_factory=list)
    status: DIDStatus = DIDStatus.ACTIVE
    created: str = ""
    updated: str = ""

    def is_active(self) -> bool:
        """Check whether this DID is active.

        Returns:
            True if status is ACTIVE.
        """
        return self.status == DIDStatus.ACTIVE

    def to_dict(self) -> dict[str, Any]:
        """Serialize the DID Document to a dictionary.

        Returns:
            JSON-serializable dictionary representation.
        """
        doc: dict[str, Any] = {
            "@context": self.context,
            "id": self.id,
            "verificationMethod": [
                {
                    "id": vm.id,
                    "type": vm.type,
                    "controller": vm.controller,
                    "publicKeyHex": vm.public_key_hex,
                }
                for vm in self.verification_methods
            ],
            "authentication": self.authentication,
            "assertionMethod": self.assertion_method,
            "service": [
                {"id": s.id, "type": s.type, "serviceEndpoint": s.endpoint}
                for s in self.services
            ],
            "status": self.status.value,
        }
        if self.created:
            doc["created"] = self.created
        if self.updated:
            doc["updated"] = self.updated
        return doc

File: sdk/go/did/document.go
// Package did implements W3C Decentralized Identifier (DID) document types
// and resolution for the ZeroID did:zero method.
package did

import "fmt"

// IdentityStatus represents the lifecycle status of a DID identity on-chain.
type IdentityStatus int

const (
	// StatusInactive indicates the identity has not been activated.
	StatusInactive IdentityStatus = iota
	// StatusActive indicates the identity is active and usable.
	StatusActive
	// StatusSuspended indicates the identity has been temporarily suspended.
	StatusSuspended
	// StatusRevoked indicates the identity has been permanently revoked.
	StatusRevoked
)

// String returns the human-readable name for an IdentityStatus.
func (s IdentityStatus) String() string {
	switch s {
	case StatusInactive:
		return "Inactive"
	case StatusActive:
		return "Active"
	case StatusSuspended:
		return "Suspended"
	case StatusRevoked:
		return "Revoked"
	default:
		return fmt.Sprintf("Unknown(%d)", int(s))
	}
}

// DIDDocument represents a W3C DID Document as defined in
// https://www.w3.org/TR/did-core/.
type DIDDocument struct {
	// Context is the JSON-LD context for the DID document.
	Context []string `json:"@context"`
	// ID is the DID URI that identifies this document.
	ID string `json:"id"`
	// Controller is the DID of the entity controlling this document.
	Controller string `json:"controller,omitempty"`
	// VerificationMethod lists the verification methods associated with this DID.
	VerificationMethod []VerificationMethod `json:"verificationMethod,omitempty"`
	// Authentication lists verification method IDs used for authentication.
	Authentication []string `json:"authentication,omitempty"`
	// AssertionMethod lists verification method IDs used for assertions.
	AssertionMethod []string `json:"assertionMethod,omitempty"`
	// Service lists service endpoints associated with this DID.
	Service []Service `json:"service,omitempty"`
	// Status is the current lifecycle status of this identity.
	Status IdentityStatus `json:"status"`
}

// VerificationMethod represents a cryptographic public key or verification
// method associated with a DID, as defined in the W3C DID Core specification.
type VerificationMethod struct {
	// ID is the unique identifier for this verification method.
	ID string `json:"id"`
	// Type is the type of verification method (e.g., "EcdsaSecp256k1VerificationKey2019").
	Type string `json:"type"`
	// Controller is the DID that controls this verification method.
	Controller string `json:"controller"`
	// PublicKeyHex is the hex-encoded public key.
	PublicKeyHex string `json:"publicKeyHex,omitempty"`
	// PublicKeyMultibase is the multibase-encoded public key.
	PublicKeyMultibase string `json:"publicKeyMultibase,omitempty"`
}

// Service represents a service endpoint associated with a DID,
// such as a messaging or credential exchange service.
type Service struct {
	// ID is the unique identifier for this service.
	ID string `json:"id"`
	// Type is the type of service (e.g., "CredentialRegistry").
	Type string `json:"type"`
	// ServiceEndpoint is the URL of the service endpoint.
	ServiceEndpoint string `json:"serviceEndpoint"`
}
